terraso_backend/apps/graphql/schema/sites.py

```diff
@@ -187,25 +195,36 @@ def mutate_and_get_payload(cls, root, info, **kwargs):
         log = cls.get_logger()
         user = info.context.user
         site = cls.get_or_throw(Site, "id", kwargs["id"])
-        if not user.has_perm(Site.get_perm("change"), site):
-            raise cls.not_allowed(MutationTypes.UPDATE)
 
-        # if any site settings fields are present in the mutuation, check that the user
-        # has the associated settings-change permission as well
-        if Site.SETTINGS_FIELDS.intersection(kwargs.keys()):
-            if not user.has_perm(Site.get_perm("change_settings"), site):
-                raise cls.not_allowed(MutationTypes.UPDATE)
+        if not check_site_permission(user, "update_settings", Context(site=site)):
+            raise cls.not_allowed(MutationTypes.UPDATE)
 
         project_id = kwargs.pop("project_id", False)
         result = super().mutate_and_get_payload(root, info, **kwargs)
         if project_id is False:
             # no project id included
             return result
+
+        # otherwise, we might need to update the site's project
         site = result.site
+
+        # determine the potential target project
         if project_id is not None:
             project = Project.objects.get(id=project_id)
-            if not user.has_perm(Project.get_perm("add_site"), project):
+        else:
+            project = None
+
+        # check if we're allowed to make the change
+        context = Context(project=project, source_site=site)
+        if site.is_unaffiliated:
+            if not check_project_permission(user, "add_unaffiliated_site", context):
                 raise cls.not_allowed(MutationTypes.UPDATE)
+        else:
+            if not check_project_permission(user, "transfer_affiliated_site", context):
+                raise cls.not_allowed(MutationTypes.UPDATE)
```

Comment:
question: for `check_site_permission`, we dispatch to the appropriate affiliated/unaffiliated permission check transparently inside `permission_matrix.py`. is there a reason not to do the same for `check_project_permission`? asking because this logic ends up getting repeated further down the file